{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Supplementary_Material.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "mount_file_id": "1_sU8Gvmzx0hPvaLF0tyo84ZFgzEfwXca",
      "authorship_tag": "ABX9TyMFY1/zuF3Dg+7Gja62hYEE",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/saulobritto/bioinfo-tools-gcolab/blob/main/Supplementary_Material.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xjfx_Z3NqzXG"
      },
      "source": [
        "\n",
        "---\n",
        "\n",
        "\n",
        "#**Whole genome sequencing of 53 Brucella abortus strains reveals variants possible associated with reduced susceptibility to rifampicin and resistance to multiple antimicrobial agents**\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "InexjcFGe0C0"
      },
      "source": [
        "##Code for PARSNP, .gbk PROKKA file and CARD Intersection\n",
        "\n",
        "    This code gets in .gbk PROKKA files the genes that match \n",
        "    the locus_tag gave by PARSNP run and intersect this \n",
        "    search with the CARD database genes list.\n",
        "\n",
        "---\n",
        "\n",
        "    Author:\n",
        "      Saulo Britto da Silva\n",
        "        saulobdasilva@gmail.com\n",
        "        https://github.com/saulobritto\n",
        "\n",
        "    > Last update: 08 Jan. 2021\n",
        "\n",
        "---\n",
        "    [BIOPYTHON]\n",
        "    Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczynski, Michiel J. L. de Hoon:\n",
        "    Biopython: freely available Python tools for computational molecular biology and bioinformatics. \n",
        "    Bioinformatics, Volume 25, Issue 11, 1 June 2009, Pages 1422–1423, https://doi.org/10.1093/bioinformatics/btp163\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "gHcQH1AO1uCy"
      },
      "source": [
        "#biopython install\n",
        "!pip install biopython"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cVj1jKX711dl"
      },
      "source": [
        "#libraries import\n",
        "import os\n",
        "from Bio.Seq import Seq\n",
        "from Bio.SeqRecord import SeqRecord\n",
        "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
        "from Bio import SeqIO\n",
        "from io import StringIO\n",
        "import pathlib\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import json\n",
        "from itertools import chain"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "SMaIKpC03mhx"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "SaGcNsWmcopS"
      },
      "source": [
        "#locus_tags file import, organized in excel with the cro1 + cro2 variants generated by parsnp\n",
        "from google.colab import files\n",
        "files.upload()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6eZ4jLA51d1T"
      },
      "source": [
        "#DataFrame created from file\n",
        "df_locus_tags = pd.read_csv('/file-imported-path', ',')\n",
        "\n",
        "#create the locus_tag list\n",
        "locus_tag_list = df_locus_tags['column-name-with-locus_tags'].tolist()\n",
        "\n",
        "print(df_locus_tags)\n",
        "print(locus_tag_list)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O7ZtRJ1N1rVZ"
      },
      "source": [
        "#path to gbk file folders of both chromosomes generated by PROKKA for the Brucella abortus ref 2308\n",
        "ref1 = list(pathlib.Path(\"/content/drive/MyDrive/IC UFLA/Referências/prokka_2308_cro1\").glob(\"*.gbk\"))\n",
        "ref2 = list(pathlib.Path(\"/content/drive/MyDrive/IC UFLA/Referências/prokka_2308_cro2\").glob(\"*.gbk\"))\n",
        "\n",
        "#join path\n",
        "ref = ref1 + ref2\n",
        "\n",
        "genes = []\n",
        "\n",
        "#search from the PARSNP locus_tag, the genes an products from the reference\n",
        "for file in ref:\n",
        "  recs = SeqIO.to_dict(SeqIO.parse(file, \"genbank\"))\n",
        "  for i in recs:\n",
        "    for f in recs[i].features:\n",
        "        if f.type == \"CDS\":\n",
        "           for z in locus_tag_list:\n",
        "            if z in f.qualifiers.get('locus_tag', []):\n",
        "                print(f)\n",
        "                #create the list with genes and proteins\n",
        "                itens = [f.qualifiers.get('gene', [])]\n",
        "                genes.extend(itens)\n",
        "\n",
        "#print the list, for checking results                \n",
        "print(genes)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sohaXg0X4PUR"
      },
      "source": [
        "#genes found DataFrame\n",
        "g_p_df = pandas.DataFrame(genes)\n",
        "g_p_df.columns =['Gene']\n",
        "print(g_p_df)\n",
        "\n",
        "#create a list with the genes\n",
        "genes_match = g_p_df['Gene'].tolist()\n",
        "\n",
        "print('\\n', 'Genes match from the intersection between PROKKA and PARSNP:', '\\n', genes_match)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ra4RfcU54uDs"
      },
      "source": [
        "#corrects the result above, splitting genes followed by a \"_1 or _2\", getting only the gene name\n",
        "genes_prokka_parsnp = []\n",
        "for i in genes_match:\n",
        "    z = str(i).split('_')\n",
        "    genes_prokka_parsnp.append(z)\n",
        "\n",
        "#corrects the result above, that found lists inside a list\n",
        "correction_list = []\n",
        "for l in genes_prokka_parsnp:\n",
        "  for i in l:\n",
        "    correction_list.append(i)\n",
        "\n",
        "genes_prokka_parsnp = correction_list\n",
        "\n",
        "print('The genes found by the intersection between PROKKA and PARSNP: ', '\\n', genes_prokka_parsnp)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nrYZKCiY7AKD"
      },
      "source": [
        "#upload the 'card.JSON' file version 3.1.0 from CARD Database \n",
        "files.upload()"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IgRhY4BL7cSS"
      },
      "source": [
        "#import and read JSON file\n",
        "file_name = '/content/card.json'\n",
        "with open(file_name, 'r') as f:\n",
        "    document =  json.loads(f.read())\n",
        "\n",
        "#create a DataFrame\n",
        "card_json = pandas.DataFrame.from_dict(document, orient='columns')\n",
        "print(card_json.head())\n",
        "\n",
        "#select the genes row itens\n",
        "genes_row = card_json.iloc[1].tolist()\n",
        "\n",
        "#some results were phrases with genes, this code below splits the phrases and we get the genes separated\n",
        "genes_row_corrected = []\n",
        "for i in genes_row:\n",
        "  z = str(i).split(' ')\n",
        "  genes_row_corrected.append(z)\n",
        "\n",
        "#normaliza os dados que ficaram em listas dentro de listas para uma lista única, para realizar a interseção\n",
        "genes_card = []\n",
        "for l in genes_row_corrected:\n",
        "  for i in l:\n",
        "    genes_card.append(i)\n",
        "print('\\n', 'This is the genes list from CARD Database:', '\\n', genes_card)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "9-1dvanI8vDO"
      },
      "source": [
        "total_intersection = list(set(genes_prokka_parsnp) & set(genes_card))\n",
        "print('These are the final genes from the intersection between PROKKA, PARSNP and CARD Database:', '\\n', total_intersection)"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}